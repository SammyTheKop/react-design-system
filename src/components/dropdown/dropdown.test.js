// react
import React, { useState } from "react";

// component
import Dropdown from "./dropdown";

// jest renderer
import renderer from "react-test-renderer";

// testing library react
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

// Dropdown List that can be reused by the Dropdown Component below in terms of executing the test cases
export const dropdownList = [
  {
    id: 1,
    name: "Dropdown Content - 1",
    clickHandlerFunc: (e, id, name) => {
      console.log("Button Clicked", id, name);
    },
  },
  {
    id: 2,
    name: "Dropdown Content - 2",
    clickHandlerFunc: (e, id, name) => {
      console.log("Button Clicked", id, name);
    },
  },
  {
    id: 3,
    name: "Dropdown Content - 3",
    clickHandlerFunc: (e, id, name) => {
      console.log("Button Clicked", id, name);
    },
  },
];

// Case 1: Dropdown Renders properly or not
it("Case 1: Dropdown Component Renders Correctly", () => {
  const dropdownComponentTree = renderer
    .create(<Dropdown dropdownName="Dropdown" dropdownList={dropdownList} />)
    .toJSON();
  expect(dropdownComponentTree).toMatchSnapshot();
});

// Case 2: Checks the Dropdown Button Test id is correct or not
it("Case 2: Checks the Dropdown Button Test id is correct or not", () => {
  const { getByTestId } = render(
    <Dropdown dropdownName="Dropdown" dropdownList={dropdownList} />
  );
  expect(getByTestId("dropdown-button")).toBeInTheDocument();
});

// Case 3: Checks the number of the dropdown buttons
it("Case 3: Checks the number of the dropdown buttons", () => {
  const { getAllByTestId } = render(
    <Dropdown dropdownName="Dropdown" dropdownList={dropdownList} />
  );
  expect(getAllByTestId("dropdown-content-button").length).toEqual(3);
});

// Case 4: Checks the dropdown name is correct or not
it("Case 4: Checks the dropdown name is correct or not", () => {
  const { getByText } = render(
    <Dropdown dropdownName="Dropdown" dropdownList={dropdownList} />
  );
  expect(getByText("Dropdown")).toBeInTheDocument();
});

/**
 * @name TestComponent
 * @details  example use case of the dropdown button component
 * @returns Dropdown button component with the updated count
 */
const TestComponent = () => {
  const [firstDropdownValue, setFirstDropdownValue] = useState(0);
  const [secondDropdownValue, setSecondDropdownValue] = useState(0);
  const [thirdDropdownValue, setThirdDropdownValue] = useState(0);

  const dropdownList = [
    {
      id: 1,
      name: `Dropdown Content - 1 | Value: ${firstDropdownValue}`,
      clickHandlerFunc: (e, id, name) => {
        setFirstDropdownValue(firstDropdownValue + 1);
      },
    },
    {
      id: 2,
      name: `Dropdown Content - 2 | Value: ${secondDropdownValue}`,
      clickHandlerFunc: (e, id, name) => {
        setSecondDropdownValue(secondDropdownValue + 1);
      },
    },
    {
      id: 3,
      name: `Dropdown Content - 3 | Value: ${thirdDropdownValue}`,
      clickHandlerFunc: (e, id, name) => {
        setThirdDropdownValue(thirdDropdownValue + 1);
      },
    },
  ];

  return <Dropdown dropdownName="Dropdown" dropdownList={dropdownList} />;
};

// Case 5: Checks the functiona
it("Case 5: Dropdown Button Click Check", () => {
  const { getByText } = render(<TestComponent />);

  // Click event using JEST for all the dropdown buttons
  fireEvent.click(getByText("Dropdown Content - 1 | Value: 0"));
  fireEvent.click(getByText("Dropdown Content - 2 | Value: 0"));
  fireEvent.click(getByText("Dropdown Content - 3 | Value: 0"));

  // validate the button click returns the correct resutl
  expect(getByText("Dropdown Content - 1 | Value: 1")).toBeInTheDocument();
  expect(getByText("Dropdown Content - 2 | Value: 1")).toBeInTheDocument();
  expect(getByText("Dropdown Content - 3 | Value: 1")).toBeInTheDocument();
});
