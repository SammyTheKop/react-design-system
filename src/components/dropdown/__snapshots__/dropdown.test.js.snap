// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Case 1: Dropdown Component Renders Correctly 1`] = `
Array [
  <div
    className="dropdown-component"
  >
    <button
      className="dropdown-button"
      data-testid="dropdown-button"
    >
      Dropdown
      <i
        className="dropdown-button-arrow-down"
      />
    </button>
    <div
      className="dropdown-content"
    >
      <button
        className="dropdown-content-buttons"
        data-testid="dropdown-content-button"
        onClick={[Function]}
        type="button"
      >
        Dropdown Content - 1
      </button>
      <style
        jsx={true}
      >
        
                  /* Links inside the dropdown */
                  .dropdown-content-buttons {
                    background: whitesmoke;
                    margin: 0px;
                    border: 0px;
                    box-shadow: none;
                    color: black;
                    padding: 16px;
                    display: block;
                    width: 100%;
                  }

                  /* Change color of dropdown links on hover */
                  .dropdown-content-buttons:hover {
                    background: wheat;
                  }
                
      </style>
      <button
        className="dropdown-content-buttons"
        data-testid="dropdown-content-button"
        onClick={[Function]}
        type="button"
      >
        Dropdown Content - 2
      </button>
      <style
        jsx={true}
      >
        
                  /* Links inside the dropdown */
                  .dropdown-content-buttons {
                    background: whitesmoke;
                    margin: 0px;
                    border: 0px;
                    box-shadow: none;
                    color: black;
                    padding: 16px;
                    display: block;
                    width: 100%;
                  }

                  /* Change color of dropdown links on hover */
                  .dropdown-content-buttons:hover {
                    background: wheat;
                  }
                
      </style>
      <button
        className="dropdown-content-buttons"
        data-testid="dropdown-content-button"
        onClick={[Function]}
        type="button"
      >
        Dropdown Content - 3
      </button>
      <style
        jsx={true}
      >
        
                  /* Links inside the dropdown */
                  .dropdown-content-buttons {
                    background: whitesmoke;
                    margin: 0px;
                    border: 0px;
                    box-shadow: none;
                    color: black;
                    padding: 16px;
                    display: block;
                    width: 100%;
                  }

                  /* Change color of dropdown links on hover */
                  .dropdown-content-buttons:hover {
                    background: wheat;
                  }
                
      </style>
    </div>
  </div>,
  <style
    jsx={true}
  >
    
          /* Styles for the Dropdown Component */
          .dropdown-component {
            display: inline-block;
            position: relative;
          }

          .dropdown-component .dropdown-button {
            background: #4caf50;
            color: white;
            font-weight: 700;
            border: none;
            cursor: pointer;
            min-width: 200px;
            border-radius: 4px;
          }

          .dropdown-component .dropdown-button-arrow-down {
            border: solid white;
            border-width: 0 3px 3px 0;
            display: inline-block;
            padding: 4px;
            transform: rotate(45deg);
            margin-left: 12px;
          }

          /* Dropdown Content (Hidden by Default) */
          .dropdown-component .dropdown-content {
            display: none;
            position: absolute;
            background: white;
            min-width: 200px;
            box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
            z-index: 1;
          }

          /* Show the dropdown menu on hover */
          .dropdown-component:hover .dropdown-content {
            display: block;
          }

          .dropdown-component .dropdown-button:hover {
            background: #3e8e41;
          }
        
  </style>,
]
`;
